### PR TITLE
Initialize AEA programmatically

### LINKED ISSUE #312: Initialize and run AEA programmatically
**Is your feature request related to a problem? Please describe.**

The only way to set up and run an agent is via command line using the tool `aea`.

**Describe the solution you'd like**

Being able to set up an `AEA` programmatically., and provide good support for this approach - including being able to run one programatically - ideally in a thread if desired (needed for working with things like TKinter which need to run in the main thread).

**Describe alternatives you've considered**

None.

**Additional context**

None.

### PR DESCRIPTION
**Pull request by marcofavorito**
## Proposed changes

This PR lets the developer to initialize the resources of an AEA programmatically.

## Fixes

Fixes #312 

## Types of changes

What types of changes does your code introduce to agents-aea?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply._

- [X] I have read the [CONTRIBUTING](../master/CONTRIBUTING.rst) doc
- [X] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [X] Lint and unit tests pass locally with my changes
- [X] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that code coverage does not decrease.
- [X] I have checked that the documentation about the `aea cli` tool works
- [ ] I have added necessary documentation (if appropriate)
- [X] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: aea/aea.py [hunk 1]
```diff
@@ -53,11 +54,12 @@ def __init__(self, name: str,
         :param mailbox: the mailbox of the agent.
         :param wallet: the wallet of the agent.
         :param ledger_apis: the ledger apis of the agent.
+        :param resources: the resources of the agent. If None, the directory parameter will be considered.
+        :param directory: the path to the agent's resource directory.
```
Comment (DavidMinarsch) [inline aea/aea.py]: Why are we adding a path if we pass in resources? This is rhetoric, as I know why :). Seems messy. I think we could do the following: instantiate `resources` and make it default requirement. But instantiate means it just has the path as part of the object constructor, no loading of resources at this point. Then give it a `load` method which actually populates the registry.

File: aea/aea.py [hunk 2]
```diff
@@ -112,8 +119,9 @@ def setup(self) -> None:
 
         :return: None
         """
-        self._resources = Resources.from_resource_dir(self._directory, self.context)
-        assert self._resources is not None, "No resources initialized. Error in setup."
+        if self._resources is None:
```
Comment (DavidMinarsch) [inline aea/aea.py]: here we'd do `self.resources.load`

File: aea/aea.py [hunk 3]
```diff
@@ -42,29 +42,27 @@ def __init__(self, name: str,
...
-                 directory: str = '') -> None:
+                 max_reactions: int = 20) -> None:
         """
         Instantiate the agent.
 
         :param name: the name of the agent
         :param mailbox: the mailbox of the agent.
         :param wallet: the wallet of the agent.
         :param ledger_apis: the ledger apis of the agent.
+        :param resources: the resources of the agent. If None, the directory parameter will be considered.
```
Comment (DavidMinarsch) [inline aea/aea.py]: Please update the docstring accordingly.

File: aea/aea.py [hunk 4]
```diff
@@ -81,7 +79,7 @@ def __init__(self, name: str,
                                      self.decision_maker.ownership_state,
                                      self.decision_maker.preferences,
                                      self.decision_maker.is_ready_to_pursuit_goals)
-        self._resources = None  # type: Optional[Resources]
+        self._resources = resources  # type: Optional[Resources]
         self._filter = None  # type: Optional[Filter]
```
Comment (DavidMinarsch) [inline aea/aea.py]: The filter can now be initialised here: `Filter(self.resources, self.decision_maker.message_out_queue)`

File: aea/aea.py [hunk 5]
```diff
@@ -112,9 +115,8 @@ def setup(self) -> None:
 
         :return: None
         """
-        self._resources = Resources.from_resource_dir(self._directory, self.context)
-        assert self._resources is not None, "No resources initialized. Error in setup."
-        self._resources.setup()
+        self.resources.load(self.context)
+        self.resources.setup()
         self._filter = Filter(self.resources, self.decision_maker.message_out_queue)
```
Comment (DavidMinarsch) [inline aea/aea.py]: We can move this initialisation up now.

File: aea/cli/run.py [hunk 6]
```diff
@@ -209,7 +210,7 @@ def run(click_context, connection_name: str, env_file: str, install_deps: bool):
             click_context.invoke(install)
 
     mailbox = MailBox(connection)
-    agent = AEA(agent_name, mailbox, wallet, ledger_apis, directory=str(Path(".")))
+    agent = AEA(agent_name, mailbox, wallet, ledger_apis, resources=Resources(str(Path("."))))
```
Comment (DavidMinarsch) [inline aea/cli/run.py]: neat!

File: aea/aea.py (hunk 4 above)
Comment (DavidMinarsch) [reply aea/aea.py]: You might also have to check the asserts in the `Filter` class

File: aea/registries/base.py [hunk 7]
```diff
@@ -468,19 +469,15 @@ def __init__(self):
 
         self._registries = [self.protocol_registry, self.handler_registry, self.behaviour_registry, self.task_registry]
 
-    @classmethod
-    def from_resource_dir(cls, directory: str, agent_context: AgentContext) -> Optional['Resources']:
```
Comment (marcofavorito) [inline aea/registries/base.py]: this was an obsolete method. The suggested approach now would be:
```python
resources = Resources(directory_path)
aea = AEA(..., resources=resources)
resources.load(aea.context)
```

File: tests/test_aea.py [hunk 8]
```diff
@@ -203,9 +202,9 @@ def setup_class(cls):
         cls.connection = DummyConnection()
         cls.mailbox = MailBox(cls.connection)
 
-        cls.aea = AEA(cls.agent_name, cls.mailbox, cls.wallet, cls.ledger_apis)
-        cls.resources = Resources.from_resource_dir(os.path.join(CUR_PATH, "data", "dummy_aea"), cls.aea.context)
-        cls.aea.resources = cls.resources
+        cls.resources = Resources(os.path.join(CUR_PATH, "data", "dummy_aea"))
+        cls.aea = AEA(cls.agent_name, cls.mailbox, cls.wallet, cls.ledger_apis, cls.resources)
+        cls.resources.load(cls.aea.context)
```
Comment (marcofavorito) [inline tests/test_aea.py]: example

File: aea/registries/base.py (hunk 7 above)
Comment (marcofavorito) [reply aea/registries/base.py]: notice, however, that `resources.load` is not mandatory.
It is called by the framework on `aea.start()`.

Comment (DavidMinarsch) [reply aea/registries/base.py]: Yes, so we should probably not add the `resources.load(area.context)` line to the demos.

File: tests/test_aea.py [hunk 9]
```diff
@@ -180,3 +187,110 @@ def test_handle():
...
+        cls.agent_name = "MyAgent"
+        cls.private_key_pem_path = os.path.join(CUR_PATH, "data", "priv.pem")
+        cls.wallet = Wallet({'default': cls.private_key_pem_path})
+        cls.ledger_apis = LedgerApis({})
+        cls.connection = DummyConnection()
+        cls.mailbox = MailBox(cls.connection)
+
+        cls.resources = Resources(os.path.join(CUR_PATH, "data", "dummy_aea"))
+        cls.aea = AEA(cls.agent_name, cls.mailbox, cls.wallet, cls.ledger_apis, cls.resources)
+        cls.resources.load(cls.aea.context)
```
Comment (DavidMinarsch) [inline tests/test_aea.py]: Maybe drop as it is not needed?

File: tests/test_aea.py [hunk 10]
```diff
@@ -180,3 +187,110 @@ def test_handle():
...
+        cls.agent_name = "MyAgent"
+        cls.private_key_pem_path = os.path.join(CUR_PATH, "data", "priv.pem")
+        cls.wallet = Wallet({'default': cls.private_key_pem_path})
+        cls.ledger_apis = LedgerApis({})
+        cls.connection = DummyConnection()
+        cls.mailbox = MailBox(cls.connection)
+
+        cls.temp = tempfile.mkdtemp(prefix="test_aea_resources")
+        cls.resources = Resources(cls.temp)
+        cls.aea = AEA(cls.agent_name, cls.mailbox, cls.wallet, cls.ledger_apis, resources=cls.resources)
```
Comment (DavidMinarsch) [inline tests/test_aea.py]: @dishmop this is now the new way to build an AEA programmatically. Any feedback?

File: aea/registries/base.py (hunk 7 above)
Comment (marcofavorito) [reply aea/registries/base.py]: yes, agreed.

File: tests/test_aea.py (hunk 9 above)
Comment (marcofavorito) [reply tests/test_aea.py]: true, I'll remove it.

**Pull request merged**